druid/src/text/text_input.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Map input to `EditAction`s
+
+use super::Movement;
+use crate::{HotKey, KeyCode, SysMods};
+use druid_shell::KeyEvent;
+
+// This following enumerations are heavily inspired by xi-editors enumerations found at
+// https://github.com/xi-editor/xi-editor/blob/e2589974fc4050beb33af82481aa71b258358e48/rust/core-lib/src/edit_types.rs
+// This is done with the goal of eventually being able to easily switch
+// to a xi-based implementation of our Events.
+
+/// Defines actions that edit the state of a textbox widget
+#[derive(Debug, PartialEq, Clone)]
+pub enum EditAction {
+    Move(Movement),
+    ModifySelection(Movement),
+    SelectAll,
+    Click(MouseAction),
+    Drag(MouseAction),
+    Delete,
+    Backspace,
+    Insert(String),
+    Paste(String),
+}
+
+/// Extra information related to mouse actions
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct MouseAction {
+    pub column: usize,
+    pub shift: bool,
```

Comment:
I think this should be row, column, and `KeyModifiers`. We do care about ctrl/cmd/etc.